### PR TITLE
Make SQL Schema Describer to understand index length and direction

### PR DESCRIPTION
**Pull request by pimeys**
Changes a few things in the describer public API:

- Primary key columns are now special structs, that can hold length and sort
  order information.
- Index columns are similar, with length and sort order.
- Needs the flag `extendedIndexes` set.

Part of: https://github.com/prisma/prisma/issues/9577

### CODE REVIEW

File: introspection-engine/connectors/sql-introspection-connector/src/introspection.rs [hunk 1]
```diff
@@ -58,16 +58,24 @@ pub fn introspect(
         }
 
         for index in &table.indices {
+            // TODO: enable with preview flag, when dml index extensions are done.
+            if index.columns.iter().any(|c| c.length.is_some()) {
```
Comment (pimeys) [inline introspection-engine/connectors/sql-introspection-connector/src/introspection.rs:62]: maybe remove

File: introspection-engine/connectors/sql-introspection-connector/src/introspection.rs [hunk 2]
```diff
@@ -58,16 +58,24 @@ pub fn introspect(
...
 
-        if let Some(pk) = &table.primary_key {
-            model.primary_key = Some(PrimaryKeyDefinition {
-                name: None,
-                db_name: pk.constraint_name.clone(),
-                fields: pk.columns.clone(),
-                defined_on_field: pk.columns.len() == 1,
-            });
+        match &table.primary_key {
+            Some(pk) if pk.columns.iter().all(|c| c.length.is_none()) => {
```
Comment (pimeys) [inline introspection-engine/connectors/sql-introspection-connector/src/introspection.rs:70]: maybe remove

File: introspection-engine/connectors/sql-introspection-connector/src/introspection.rs (hunk 1 above)
Comment (pimeys) [reply introspection-engine/connectors/sql-introspection-connector/src/introspection.rs:62]: This one now is needed, we do not filter them in the describer anymore...

File: introspection-engine/connectors/sql-introspection-connector/src/introspection.rs (hunk 2 above)
Comment (pimeys) [reply introspection-engine/connectors/sql-introspection-connector/src/introspection.rs:70]: This one now is needed, we do not filter them in the describer anymore...

File: libs/sql-schema-describer/Cargo.toml [hunk 3]
```diff
@@ -7,6 +7,7 @@ version = "0.1.0"
 [dependencies]
 native-types = { path = "../native-types" }
 prisma-value = { path = "../prisma-value" }
+datamodel = { path = "../datamodel/core" }
```
Comment (tomhoule) [inline libs/sql-schema-describer/Cargo.toml:10]: (haven't read the rest of the PR yet) — why datamodel here? I'm not comfortable with this.

Comment (pimeys) [reply libs/sql-schema-describer/Cargo.toml]: no need. taken out.

**Review by tomhoule — APPROVED** [review]

**Pull request merged**